File: backend/app/models/tesoreria.py
"""
Modelos para el módulo de Tesorería
Gestiona cheques, movimientos de caja, cuentas bancarias y flujos de fondos
"""
from sqlalchemy import Column, String, Numeric, Text, Date, DateTime, ForeignKey, Boolean, Integer, Index
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
import enum

from app.models.base import BaseModel, get_argentina_now


# ==================== ENUMS ====================

class TipoChequeEnum(str, enum.Enum):
    """Tipo de cheque"""
    FISICO = "fisico"
    ECHEQ = "echeq"


class OrigenChequeEnum(str, enum.Enum):
    """Origen del cheque"""
    RECIBIDO_CLIENTE = "recibido_cliente"  # Cheque recibido de un cliente como pago
    RECIBIDO_PROVEEDOR = "recibido_proveedor"  # Cheque recibido de un proveedor (ej: devolución)
    EMITIDO = "emitido"  # Cheque emitido por la empresa


class EstadoChequeEnum(str, enum.Enum):
    """Estado del cheque"""
    EN_CARTERA = "en_cartera"  # Cheque en poder de la empresa
    DEPOSITADO = "depositado"  # Depositado en banco, esperando acreditación
    COBRADO = "cobrado"  # Cobrado/acreditado en cuenta
    ENTREGADO = "entregado"  # Entregado a tercero como forma de pago
    RECHAZADO = "rechazado"  # Rechazado por el banco
    ANULADO = "anulado"  # Anulado del sistema


class TipoMovimientoTesoreriaEnum(str, enum.Enum):
    """Tipo de movimiento de tesorería"""
    INGRESO_EFECTIVO = "ingreso_efectivo"
    INGRESO_TRANSFERENCIA = "ingreso_transferencia"
    INGRESO_CHEQUE = "ingreso_cheque"
    EGRESO_EFECTIVO = "egreso_efectivo"
    EGRESO_TRANSFERENCIA = "egreso_transferencia"
    EGRESO_CHEQUE = "egreso_cheque"
    DEPOSITO_CHEQUE = "deposito_cheque"
    COBRO_CHEQUE = "cobro_cheque"


class MetodoPagoEnum(str, enum.Enum):
    """Método de pago"""
    EFECTIVO = "efectivo"
    TRANSFERENCIA = "transferencia"
    CHEQUE = "cheque"
    E_CHEQUE = "e_cheque"


# Lista de bancos de Argentina
BANCOS_ARGENTINA = [
    "Banco Nación",
    "Banco Provincia",
    "Banco Ciudad",
    "Banco Santander",
    "Banco Galicia",
    "Banco BBVA",
    "Banco Macro",
    "Banco HSBC",
    "Banco Credicoop",
    "Banco Patagonia",
    "Banco Supervielle",
    "Banco ICBC",
    "Banco Hipotecario",
    "Otro"
]


# ==================== MODELOS ====================

class Cheque(BaseModel):
    """
    Gestión de cheques (físicos y e-cheqs)
    """

    __tablename__ = "cheques"

    # Identificación del cheque
    numero = Column(String(50), nullable=False, index=True)
    tipo = Column(String(20), default=TipoChequeEnum.FISICO.value, nullable=False)
    origen = Column(String(30), default=OrigenChequeEnum.RECIBIDO_CLIENTE.value, nullable=False)
    estado = Column(String(20), default=EstadoChequeEnum.EN_CARTERA.value, nullable=False, index=True)

    # Monto y fechas
    monto = Column(Numeric(12, 2), nullable=False)
    fecha_emision = Column(Date, nullable=True)
    fecha_vencimiento = Column(Date, nullable=False, index=True)
    fecha_cobro = Column(Date, nullable=True)
    fecha_deposito = Column(Date, nullable=True)
    fecha_entrega = Column(Date, nullable=True)

    # Bancos
    banco_origen = Column(String(100))  # Banco que emite el cheque
    cuenta_destino_id = Column(UUID(as_uuid=True), ForeignKey("cuentas_bancarias.id"), nullable=True)
    banco_destino = Column(String(100))  # Banco donde se deposita

    # Relaciones con entidades
    empresa_id = Column(UUID(as_uuid=True), ForeignKey("empresas.id"), nullable=True, index=True)  # Cliente si es recibido

    # Información del cheque
    librador = Column(String(200))  # Persona/empresa que emite el cheque
    cuit_librador = Column(String(20))

    # Auditoría y control
    registrado_por_id = Column(UUID(as_uuid=True), ForeignKey("usuarios.id"), nullable=False)
    cobrado_por_id = Column(UUID(as_uuid=True), ForeignKey("usuarios.id"), nullable=True)
    fecha_registro = Column(DateTime, default=get_argentina_now, nullable=False)

    # Observaciones
    notas = Column(Text)
    motivo_rechazo = Column(Text)
    concepto_entrega = Column(String(500))  # Para qué se entregó el cheque

    # Estado lógico
    activo = Column(Boolean, default=True, nullable=False)

    # Relaciones
    empresa = relationship("Empresa")
    cuenta_destino = relationship("CuentaBancaria")
    registrado_por = relationship("Usuario", foreign_keys=[registrado_por_id])
    cobrado_por = relationship("Usuario", foreign_keys=[cobrado_por_id])

    # Índices compuestos
    __table_args__ = (
        Index('ix_cheques_estado_vencimiento', 'estado', 'fecha_vencimiento'),
        Index('ix_cheques_empresa_estado', 'empresa_id', 'estado'),
    )

    def __repr__(self):
        return f"<Cheque #{self.numero} ${self.monto} - {self.estado}>"


class MovimientoTesoreria(BaseModel):
    """
    Movimientos de tesorería (ingresos y egresos de caja)
    Registra efectivo, transferencias y cheques
    """

    __tablename__ = "movimientos_tesoreria"

    # Tipo y concepto
    tipo = Column(String(30), nullable=False, index=True)
    concepto = Column(String(200), nullable=False)
    descripcion = Column(Text)

    # Monto
    monto = Column(Numeric(12, 2), nullable=False)
    es_ingreso = Column(Boolean, nullable=False)

    # Fechas
    fecha_movimiento = Column(Date, nullable=False, index=True)
    fecha_valor = Column(Date, nullable=True)  # Para cheques diferidos

    # Método de pago
    metodo_pago = Column(String(20), nullable=False)

    # Si es transferencia
    banco_origen = Column(String(100))
    banco_destino = Column(String(100))
    cuenta_destino_id = Column(UUID(as_uuid=True), ForeignKey("cuentas_bancarias.id"), nullable=True)
    numero_transferencia = Column(String(100))

    # Si es cheque
    cheque_id = Column(UUID(as_uuid=True), ForeignKey("cheques.id"), nullable=True)

    # Relaciones con entidades
    empresa_id = Column(UUID(as_uuid=True), ForeignKey("empresas.id"), nullable=True, index=True)  # Cliente o proveedor

    # Vinculación con pesajes (para cobranzas)
    pesaje_id = Column(UUID(as_uuid=True), ForeignKey("pesajes.id"), nullable=True)
    cobro_id = Column(UUID(as_uuid=True), ForeignKey("cobros_cliente.id"), nullable=True)

    # Auditoría
    registrado_por_id = Column(UUID(as_uuid=True), ForeignKey("usuarios.id"), nullable=False)
    notas = Column(Text)
    comprobante = Column(String(100))  # Referencia de comprobante

    # Anulación
    anulado = Column(Boolean, default=False, nullable=False)
    motivo_anulacion = Column(Text)
    anulado_por_id = Column(UUID(as_uuid=True), ForeignKey("usuarios.id"), nullable=True)
    fecha_anulacion = Column(DateTime, nullable=True)

    # Estado lógico
    activo = Column(Boolean, default=True, nullable=False)

    # Relaciones
    empresa = relationship("Empresa")
    cuenta_destino = relationship("CuentaBancaria")
    cheque = relationship("Cheque")
    pesaje = relationship("Pesaje")
    cobro = relationship("CobroCliente")
    registrado_por = relationship("Usuario", foreign_keys=[registrado_por_id])
    anulado_por = relationship("Usuario", foreign_keys=[anulado_por_id])

    # Índices compuestos
    __table_args__ = (
        Index('ix_mov_tesoreria_fecha_tipo', 'fecha_movimiento', 'tipo'),
        Index('ix_mov_tesoreria_empresa_fecha', 'empresa_id', 'fecha_movimiento'),
    )

    def __repr__(self):
        return f"<MovimientoTesoreria {self.tipo} ${self.monto}>"


class CajaEfectivo(BaseModel):
    """
    Representa una caja de efectivo (ej: Caja Principal, Caja Chica)
    Permite tener múltiples cajas de efectivo
    """

    __tablename__ = "cajas_efectivo"

    nombre = Column(String(100), nullable=False)
    descripcion = Column(Text)
    saldo_actual = Column(Numeric(12, 2), default=0, nullable=False)
    saldo_inicial = Column(Numeric(12, 2), default=0, nullable=False)
    moneda = Column(String(10), default="ARS", nullable=False)
    activo = Column(Boolean, default=True, nullable=False)
    es_principal = Column(Boolean, default=False, nullable=False)
    notas = Column(Text)

    def __repr__(self):
        return f"<CajaEfectivo {self.nombre} ${self.saldo_actual}>"


class MovimientoCaja(BaseModel):
    """
    Movimientos de una caja de efectivo específica
    """

    __tablename__ = "movimientos_caja"

    caja_id = Column(UUID(as_uuid=True), ForeignKey("cajas_efectivo.id"), nullable=False, index=True)

    # Tipo de movimiento
    tipo = Column(String(20), nullable=False)  # ingreso, egreso
    concepto = Column(String(200), nullable=False)
    descripcion = Column(Text)

    # Montos
    monto = Column(Numeric(12, 2), nullable=False)
    saldo_anterior = Column(Numeric(12, 2), nullable=False)
    saldo_posterior = Column(Numeric(12, 2), nullable=False)

    # Fecha
    fecha = Column(Date, nullable=False, index=True)

    # Referencias
    movimiento_tesoreria_id = Column(UUID(as_uuid=True), ForeignKey("movimientos_tesoreria.id"), nullable=True)
    empresa_id = Column(UUID(as_uuid=True), ForeignKey("empresas.id"), nullable=True)

    # Auditoría
    registrado_por_id = Column(UUID(as_uuid=True), ForeignKey("usuarios.id"), nullable=False)
    notas = Column(Text)

    # Anulación
    anulado = Column(Boolean, default=False, nullable=False)

    # Relaciones
    caja = relationship("CajaEfectivo")
    movimiento_tesoreria = relationship("MovimientoTesoreria")
    empresa = relationship("Empresa")
    registrado_por = relationship("Usuario")

    def __repr__(self):
        return f"<MovimientoCaja {self.tipo} ${self.monto}>"


class MovimientoBancario(BaseModel):
    """
    Movimientos de una cuenta bancaria específica
    """

    __tablename__ = "movimientos_bancarios"

    cuenta_id = Column(UUID(as_uuid=True), ForeignKey("cuentas_bancarias.id"), nullable=False, index=True)

    # Tipo de movimiento
    tipo = Column(String(20), nullable=False)  # ingreso, egreso
    concepto = Column(String(200), nullable=False)
    descripcion = Column(Text)

    # Montos
    monto = Column(Numeric(12, 2), nullable=False)
    saldo_anterior = Column(Numeric(12, 2), nullable=False)
    saldo_posterior = Column(Numeric(12, 2), nullable=False)

    # Fecha
    fecha = Column(Date, nullable=False, index=True)
    fecha_valor = Column(Date, nullable=True)  # Fecha de acreditación

    # Referencias
    numero_operacion = Column(String(100))
    movimiento_tesoreria_id = Column(UUID(as_uuid=True), ForeignKey("movimientos_tesoreria.id"), nullable=True)
    cheque_id = Column(UUID(as_uuid=True), ForeignKey("cheques.id"), nullable=True)
    empresa_id = Column(UUID(as_uuid=True), ForeignKey("empresas.id"), nullable=True)

    # Auditoría
    registrado_por_id = Column(UUID(as_uuid=True), ForeignKey("usuarios.id"), nullable=False)
    notas = Column(Text)

    # Anulación
    anulado = Column(Boolean, default=False, nullable=False)

    # Relaciones
    cuenta = relationship("CuentaBancaria")
    movimiento_tesoreria = relationship("MovimientoTesoreria")
    cheque = relationship("Cheque")
    empresa = relationship("Empresa")
    registrado_por = relationship("Usuario")

    def __repr__(self):
        return f"<MovimientoBancario {self.tipo} ${self.monto}>"


class Gasto(BaseModel):
    """
    Registro de gastos de la empresa
    """

    __tablename__ = "gastos"

    # Datos del gasto
    fecha = Column(Date, nullable=False, index=True)
    concepto = Column(String(200), nullable=False)
    descripcion = Column(Text)
    categoria = Column(String(100))

    # Monto
    monto = Column(Numeric(12, 2), nullable=False)

    # Proveedor (opcional)
    empresa_id = Column(UUID(as_uuid=True), ForeignKey("empresas.id"), nullable=True)
    proveedor_nombre = Column(String(200))  # Si no está en el sistema

    # Comprobante
    tipo_comprobante = Column(String(50))  # factura, ticket, recibo
    numero_comprobante = Column(String(100))
    comprobante_url = Column(String(500))

    # Método de pago
    metodo_pago = Column(String(20), nullable=False)
    cuenta_bancaria_id = Column(UUID(as_uuid=True), ForeignKey("cuentas_bancarias.id"), nullable=True)
    caja_id = Column(UUID(as_uuid=True), ForeignKey("cajas_efectivo.id"), nullable=True)
    cheque_id = Column(UUID(as_uuid=True), ForeignKey("cheques.id"), nullable=True)

    # Vinculación con otros módulos
    camion_id = Column(UUID(as_uuid=True), ForeignKey("camiones.id"), nullable=True)
    servicio_id = Column(UUID(as_uuid=True), ForeignKey("servicios.id"), nullable=True)

    # Movimiento de tesorería generado
    movimiento_tesoreria_id = Column(UUID(as_uuid=True), ForeignKey("movimientos_tesoreria.id"), nullable=True)

    # Auditoría
    registrado_por_id = Column(UUID(as_uuid=True), ForeignKey("usuarios.id"), nullable=False)
    notas = Column(Text)

    # Estado
    estado = Column(String(20), default="completado")  # completado, pendiente, anulado
    anulado = Column(Boolean, default=False, nullable=False)
    activo = Column(Boolean, default=True, nullable=False)

    # Relaciones
    empresa = relationship("Empresa")
    cuenta_bancaria = relationship("CuentaBancaria")
    caja = relationship("CajaEfectivo")
    cheque = relationship("Cheque")
    camion = relationship("Camion")
    movimiento_tesoreria = relationship("MovimientoTesoreria")
    registrado_por = relationship("Usuario")

    def __repr__(self):
        return f"<Gasto {self.concepto} ${self.monto}>"


class Recibo(BaseModel):
    """
    Recibos generados por pagos de clientes
    """

    __tablename__ = "recibos"

    # Numeración
    numero_recibo = Column(String(50), unique=True, nullable=False, index=True)

    # Fecha
    fecha = Column(Date, nullable=False, index=True)

    # Cliente
    empresa_id = Column(UUID(as_uuid=True), ForeignKey("empresas.id"), nullable=False)

    # Monto
    monto = Column(Numeric(12, 2), nullable=False)

    # Concepto
    concepto = Column(String(500), nullable=False)

    # Método de pago
    metodo_pago = Column(String(20), nullable=False)

    # Referencias
    cheque_id = Column(UUID(as_uuid=True), ForeignKey("cheques.id"), nullable=True)
    movimiento_tesoreria_id = Column(UUID(as_uuid=True), ForeignKey("movimientos_tesoreria.id"), nullable=True)
    cobro_id = Column(UUID(as_uuid=True), ForeignKey("cobros_cliente.id"), nullable=True)

    # PDF
    pdf_url = Column(String(500))

    # Auditoría
    registrado_por_id = Column(UUID(as_uuid=True), ForeignKey("usuarios.id"), nullable=False)
    notas = Column(Text)

    # Estado
    anulado = Column(Boolean, default=False, nullable=False)
    activo = Column(Boolean, default=True, nullable=False)

    # Relaciones
    empresa = relationship("Empresa")
    cheque = relationship("Cheque")
    movimiento_tesoreria = relationship("MovimientoTesoreria")
    cobro = relationship("CobroCliente")
    registrado_por = relationship("Usuario")

    def __repr__(self):
        return f"<Recibo #{self.numero_recibo} ${self.monto}>"

File: frontend/src/types/tesoreria.ts
/**
 * Tipos TypeScript para el módulo de Tesorería
 */

// ==================== CONSTANTES ====================

export const TIPOS_CHEQUE = [
  { value: 'fisico', label: 'Cheque Físico' },
  { value: 'echeq', label: 'E-Cheq' },
] as const

export const ORIGENES_CHEQUE = [
  { value: 'recibido_cliente', label: 'Recibido de Cliente' },
  { value: 'recibido_proveedor', label: 'Recibido de Proveedor' },
  { value: 'emitido', label: 'Emitido (Propio)' },
] as const

export const ESTADOS_CHEQUE = [
  { value: 'en_cartera', label: 'En Cartera', color: 'blue' },
  { value: 'depositado', label: 'Depositado', color: 'yellow' },
  { value: 'cobrado', label: 'Cobrado', color: 'green' },
  { value: 'entregado', label: 'Entregado', color: 'purple' },
  { value: 'rechazado', label: 'Rechazado', color: 'red' },
  { value: 'anulado', label: 'Anulado', color: 'gray' },
] as const

export const METODOS_PAGO = [
  { value: 'efectivo', label: 'Efectivo' },
  { value: 'transferencia', label: 'Transferencia' },
  { value: 'cheque', label: 'Cheque' },
] as const

export const BANCOS_ARGENTINA = [
  { value: 'banco_nacion', label: 'Banco Nación' },
  { value: 'banco_provincia', label: 'Banco Provincia' },
  { value: 'banco_ciudad', label: 'Banco Ciudad' },
  { value: 'banco_santander', label: 'Banco Santander' },
  { value: 'banco_galicia', label: 'Banco Galicia' },
  { value: 'banco_bbva', label: 'Banco BBVA' },
  { value: 'banco_macro', label: 'Banco Macro' },
  { value: 'banco_hsbc', label: 'Banco HSBC' },
  { value: 'banco_credicoop', label: 'Banco Credicoop' },
  { value: 'banco_patagonia', label: 'Banco Patagonia' },
  { value: 'banco_supervielle', label: 'Banco Supervielle' },
  { value: 'banco_icbc', label: 'Banco ICBC' },
  { value: 'banco_hipotecario', label: 'Banco Hipotecario' },
  { value: 'otro', label: 'Otro' },
] as const

export const CATEGORIAS_GASTO = [
  { value: 'combustible', label: 'Combustible' },
  { value: 'repuestos', label: 'Repuestos' },
  { value: 'mantenimiento', label: 'Mantenimiento' },
  { value: 'sueldos', label: 'Sueldos' },
  { value: 'servicios', label: 'Servicios' },
  { value: 'impuestos', label: 'Impuestos' },
  { value: 'alquileres', label: 'Alquileres' },
  { value: 'seguros', label: 'Seguros' },
  { value: 'administrativos', label: 'Gastos Administrativos' },
  { value: 'otros', label: 'Otros' },
] as const

// ==================== TIPOS DE ENUM ====================

export type TipoCheque = typeof TIPOS_CHEQUE[number]['value']
export type OrigenCheque = typeof ORIGENES_CHEQUE[number]['value']
export type EstadoCheque = typeof ESTADOS_CHEQUE[number]['value']
export type MetodoPago = typeof METODOS_PAGO[number]['value']
export type CategoriaGasto = typeof CATEGORIAS_GASTO[number]['value']

// ==================== CHEQUE ====================

export interface Cheque {
  id: string
  numero: string
  tipo: TipoCheque
  origen: OrigenCheque
  estado: EstadoCheque
  monto: number
  fecha_emision?: string
  fecha_vencimiento: string
  fecha_cobro?: string
  fecha_deposito?: string
  fecha_entrega?: string
  banco_origen?: string
  cuenta_destino_id?: string
  banco_destino?: string
  empresa_id?: string
  librador?: string
  cuit_librador?: string
  notas?: string
  motivo_rechazo?: string
  concepto_entrega?: string
  registrado_por_id: string
  cobrado_por_id?: string
  fecha_registro: string
  activo: boolean
  created_at: string
  updated_at: string
  // Campos enriquecidos
  empresa_nombre?: string
  registrado_por_nombre?: string
  cobrado_por_nombre?: string
  cuenta_destino_nombre?: string
  dias_para_vencimiento?: number
}

export interface ChequeCreate {
  numero: string
  tipo?: TipoCheque
  origen?: OrigenCheque
  monto: number
  fecha_emision?: string
  fecha_vencimiento: string
  banco_origen?: string
  cuenta_destino_id?: string
  banco_destino?: string
  empresa_id?: string
  librador?: string
  cuit_librador?: string
  notas?: string
}

export interface ChequeUpdate {
  numero?: string
  tipo?: TipoCheque
  origen?: OrigenCheque
  estado?: EstadoCheque
  monto?: number
  fecha_emision?: string
  fecha_vencimiento?: string
  banco_origen?: string
  cuenta_destino_id?: string
  banco_destino?: string
  empresa_id?: string
  librador?: string
  cuit_librador?: string
  notas?: string
}

export interface ChequeList {
  id: string
  numero: string
  tipo: TipoCheque
  origen: OrigenCheque
  estado: EstadoCheque
  monto: number
  fecha_vencimiento: string
  fecha_cobro?: string
  banco_origen?: string
  empresa_id?: string
  empresa_nombre?: string
  librador?: string
  dias_para_vencimiento?: number
  created_at: string
}

// Acciones de cheque
export interface DepositarChequeRequest {
  cuenta_destino_id: string
  fecha_deposito?: string
  notas?: string
}

export interface CobrarChequeRequest {
  fecha_cobro?: string
  notas?: string
}

export interface RechazarChequeRequest {
  motivo_rechazo: string
  fecha_rechazo?: string
}

export interface EntregarChequeRequest {
  empresa_destino_id?: string
  concepto: string
  fecha_entrega?: string
  notas?: string
}

// ==================== MOVIMIENTO TESORERÍA ====================

export interface MovimientoTesoreria {
  id: string
  tipo: string
  concepto: string
  descripcion?: string
  monto: number
  es_ingreso: boolean
  fecha_movimiento: string
  fecha_valor?: string
  metodo_pago: MetodoPago
  banco_origen?: string
  banco_destino?: string
  cuenta_destino_id?: string
  numero_transferencia?: string
  cheque_id?: string
  empresa_id?: string
  pesaje_id?: string
  cobro_id?: string
  notas?: string
  comprobante?: string
  registrado_por_id: string
  anulado: boolean
  motivo_anulacion?: string
  anulado_por_id?: string
  fecha_anulacion?: string
  activo: boolean
  created_at: string
  updated_at: string
  // Campos enriquecidos
  empresa_nombre?: string
  registrado_por_nombre?: string
  cheque_numero?: string
  pesaje_numero?: number
}

export interface MovimientoTesoreriaCreate {
  tipo: string
  concepto: string
  descripcion?: string
  monto: number
  es_ingreso: boolean
  fecha_movimiento: string
  fecha_valor?: string
  metodo_pago: MetodoPago
  banco_origen?: string
  banco_destino?: string
  cuenta_destino_id?: string
  numero_transferencia?: string
  cheque_id?: string
  empresa_id?: string
  pesaje_id?: string
  cobro_id?: string
  notas?: string
  comprobante?: string
}

export interface MovimientoTesoreriaList {
  id: string
  tipo: string
  concepto: string
  monto: number
  es_ingreso: boolean
  fecha_movimiento: string
  metodo_pago: MetodoPago
  empresa_id?: string
  empresa_nombre?: string
  anulado: boolean
  created_at: string
}

export interface AnularMovimientoRequest {
  motivo: string
}

// ==================== CAJA EFECTIVO ====================

export interface CajaEfectivo {
  id: string
  nombre: string
  descripcion?: string
  saldo_actual: number
  saldo_inicial: number
  moneda: string
  activo: boolean
  es_principal: boolean
  notas?: string
  created_at: string
  updated_at: string
}

export interface CajaEfectivoCreate {
  nombre: string
  descripcion?: string
  saldo_inicial?: number
  moneda?: string
  es_principal?: boolean
  notas?: string
}

export interface CajaEfectivoUpdate {
  nombre?: string
  descripcion?: string
  es_principal?: boolean
  notas?: string
  activo?: boolean
}

// ==================== MOVIMIENTO CAJA ====================

export interface MovimientoCaja {
  id: string
  caja_id: string
  tipo: 'ingreso' | 'egreso'
  concepto: string
  descripcion?: string
  monto: number
  saldo_anterior: number
  saldo_posterior: number
  fecha: string
  movimiento_tesoreria_id?: string
  empresa_id?: string
  registrado_por_id: string
  anulado: boolean
  notas?: string
  created_at: string
  updated_at: string
  // Campos enriquecidos
  caja_nombre?: string
  empresa_nombre?: string
  registrado_por_nombre?: string
}

export interface MovimientoCajaCreate {
  caja_id: string
  tipo: 'ingreso' | 'egreso'
  concepto: string
  descripcion?: string
  monto: number
  fecha: string
  empresa_id?: string
  notas?: string
}

// ==================== CUENTA BANCARIA (extendido) ====================

export interface CuentaBancariaTesoreria {
  id: string
  nombre: string
  banco: string
  tipo_cuenta?: string
  numero_cuenta?: string
  cbu?: string
  alias?: string
  titular?: string
  cuit_titular?: string
  saldo_inicial: number
  saldo_actual: number
  moneda: string
  activo: boolean
  notas?: string
  created_at: string
  updated_at: string
}

export interface CuentaBancariaCreate {
  nombre: string
  banco: string
  tipo_cuenta?: string
  numero_cuenta?: string
  cbu?: string
  alias?: string
  titular?: string
  cuit_titular?: string
  saldo_inicial?: number
  moneda?: string
  notas?: string
}

export interface CuentaBancariaUpdate {
  nombre?: string
  banco?: string
  tipo_cuenta?: string
  numero_cuenta?: string
  cbu?: string
  alias?: string
  titular?: string
  cuit_titular?: string
  notas?: string
  activo?: boolean
}

// ==================== MOVIMIENTO BANCARIO ====================

export interface MovimientoBancario {
  id: string
  cuenta_id: string
  tipo: 'ingreso' | 'egreso'
  concepto: string
  descripcion?: string
  monto: number
  saldo_anterior: number
  saldo_posterior: number
  fecha: string
  fecha_valor?: string
  numero_operacion?: string
  movimiento_tesoreria_id?: string
  cheque_id?: string
  empresa_id?: string
  registrado_por_id: string
  anulado: boolean
  notas?: string
  created_at: string
  updated_at: string
  // Campos enriquecidos
  cuenta_nombre?: string
  cuenta_banco?: string
  empresa_nombre?: string
  registrado_por_nombre?: string
}

export interface MovimientoBancarioCreate {
  cuenta_id: string
  tipo: 'ingreso' | 'egreso'
  concepto: string
  descripcion?: string
  monto: number
  fecha: string
  fecha_valor?: string
  numero_operacion?: string
  empresa_id?: string
  notas?: string
}

// ==================== GASTO ====================

export interface Gasto {
  id: string
  fecha: string
  concepto: string
  descripcion?: string
  categoria?: CategoriaGasto
  monto: number
  empresa_id?: string
  proveedor_nombre?: string
  tipo_comprobante?: string
  numero_comprobante?: string
  comprobante_url?: string
  metodo_pago: MetodoPago
  cuenta_bancaria_id?: string
  caja_id?: string
  cheque_id?: string
  camion_id?: string
  servicio_id?: string
  movimiento_tesoreria_id?: string
  registrado_por_id: string
  notas?: string
  estado: string
  anulado: boolean
  activo: boolean
  created_at: string
  updated_at: string
  // Campos enriquecidos
  empresa_nombre?: string
  cuenta_bancaria_nombre?: string
  caja_nombre?: string
  camion_identificador?: string
  registrado_por_nombre?: string
}

export interface GastoCreate {
  fecha: string
  concepto: string
  descripcion?: string
  categoria?: CategoriaGasto
  monto: number
  empresa_id?: string
  proveedor_nombre?: string
  tipo_comprobante?: string
  numero_comprobante?: string
  comprobante_url?: string
  metodo_pago: MetodoPago
  cuenta_bancaria_id?: string
  caja_id?: string
  cheque_id?: string
  camion_id?: string
  servicio_id?: string
  notas?: string
}

export interface GastoUpdate {
  fecha?: string
  concepto?: string
  descripcion?: string
  categoria?: CategoriaGasto
  monto?: number
  empresa_id?: string
  proveedor_nombre?: string
  tipo_comprobante?: string
  numero_comprobante?: string
  comprobante_url?: string
  notas?: string
}

export interface GastoList {
  id: string
  fecha: string
  concepto: string
  categoria?: CategoriaGasto
  monto: number
  metodo_pago: MetodoPago
  empresa_nombre?: string
  proveedor_nombre?: string
  estado: string
  anulado: boolean
  created_at: string
}

// ==================== RECIBO ====================

export interface Recibo {
  id: string
  numero_recibo: string
  fecha: string
  empresa_id: string
  monto: number
  concepto: string
  metodo_pago: MetodoPago
  cheque_id?: string
  movimiento_tesoreria_id?: string
  cobro_id?: string
  pdf_url?: string
  registrado_por_id: string
  notas?: string
  anulado: boolean
  activo: boolean
  created_at: string
  updated_at: string
  // Campos enriquecidos
  empresa_nombre?: string
  registrado_por_nombre?: string
}

export interface ReciboCreate {
  fecha: string
  empresa_id: string
  monto: number
  concepto: string
  metodo_pago: MetodoPago
  cheque_id?: string
  cobro_id?: string
  notas?: string
}

// ==================== RESUMEN ====================

export interface ResumenTesoreria {
  cheques_en_cartera: number
  total_cheques_cartera: number
  cheques_proximos_vencer: number
  total_proximos_vencer: number
  cheques_vencidos: number
  total_vencidos: number
  total_ingresos_efectivo: number
  total_ingresos_transferencia: number
  total_ingresos_cheque: number
  total_ingresos: number
  total_egresos_efectivo: number
  total_egresos_transferencia: number
  total_egresos_cheque: number
  total_egresos: number
  saldo_periodo: number
  saldo_cajas: number
  saldo_bancos: number
  total_disponible: number
}

// ==================== MOVIMIENTOS CONSOLIDADOS ====================

export interface MovimientoConsolidado {
  id: string
  fecha: string
  tipo: string
  origen: 'cheque' | 'movimiento_tesoreria' | 'movimiento_caja' | 'movimiento_bancario'
  concepto: string
  monto: number
  es_ingreso: boolean
  metodo_pago: MetodoPago
  empresa_nombre?: string
  numero_referencia?: string
  banco?: string
  estado?: EstadoCheque
  created_at: string
}

export interface MovimientosConsolidadosResponse {
  items: MovimientoConsolidado[]
  total: number
  total_ingresos: number
  total_egresos: number
  skip: number
  limit: number
}

// ==================== PAGINACIÓN ====================

export interface PaginatedCheques {
  items: ChequeList[]
  total: number
  skip: number
  limit: number
}

export interface PaginatedMovimientos {
  items: MovimientoTesoreriaList[]
  total: number
  skip: number
  limit: number
}

export interface PaginatedGastos {
  items: GastoList[]
  total: number
  skip: number
  limit: number
}

// ==================== FILTROS ====================

export interface FiltrosCheques {
  skip?: number
  limit?: number
  estado?: EstadoCheque | string
  tipo?: TipoCheque | string
  origen?: OrigenCheque | string
  empresa_id?: string
  fecha_desde?: string
  fecha_hasta?: string
  buscar?: string
  solo_en_cartera?: boolean
  vencidos?: boolean
  proximos_vencer?: boolean
  dias_vencimiento?: number
}

export interface FiltrosMovimientos {
  skip?: number
  limit?: number
  tipo?: string
  es_ingreso?: boolean
  metodo_pago?: MetodoPago | string
  empresa_id?: string
  fecha_desde?: string
  fecha_hasta?: string
  buscar?: string
  incluir_anulados?: boolean
}

export interface FiltrosGastos {
  skip?: number
  limit?: number
  categoria?: CategoriaGasto | string
  metodo_pago?: MetodoPago | string
  empresa_id?: string
  fecha_desde?: string
  fecha_hasta?: string
  buscar?: string
  incluir_anulados?: boolean
}

// ==================== UTILIDADES ====================

export function getEstadoChequeColor(estado: EstadoCheque): string {
  const estadoInfo = ESTADOS_CHEQUE.find(e => e.value === estado)
  return estadoInfo?.color || 'gray'
}

export function getEstadoChequeLabel(estado: EstadoCheque): string {
  const estadoInfo = ESTADOS_CHEQUE.find(e => e.value === estado)
  return estadoInfo?.label || estado
}

export function formatMonto(monto: number): string {
  return new Intl.NumberFormat('es-AR', {
    style: 'currency',
    currency: 'ARS',
  }).format(monto)
}

export function getDiasVencimiento(fechaVencimiento: string): number {
  const hoy = new Date()
  const vencimiento = new Date(fechaVencimiento)
  const diffTime = vencimiento.getTime() - hoy.getTime()
  return Math.ceil(diffTime / (1000 * 60 * 60 * 24))
}

export function getChequeVencimientoStatus(diasParaVencer?: number): {
  label: string
  color: string
} {
  if (diasParaVencer === undefined) {
    return { label: 'Sin fecha', color: 'gray' }
  }
  if (diasParaVencer < 0) {
    return { label: 'Vencido', color: 'red' }
  }
  if (diasParaVencer <= 7) {
    return { label: `${diasParaVencer} días`, color: 'yellow' }
  }
  return { label: `${diasParaVencer} días`, color: 'green' }
}
